/**:
  ros__parameters:
    
    # topics and frames
    safe_drive_topic: "safe_drive"
    real_drive_topic: "drive"
    emergency_brake_topic: "emergency_brake"
    joy_topic: "joy"
    odom_topic: "odom"
    pozyx_namespace: "pozyx"
    scan_topic: "scan"
    processed_scan_topic: "scan_processed"
    course_data_topic: "course_data"
    left_wall_pc_topic: "left_wall_pc"
    right_wall_pc_topic: "right_wall_pc"
    neighbor_info_topic: "neighbor_info"
    veh_frame: "base_link"
    lidar_frame: "laser"
    lateral_control_topic: "lateral_control"
    longitudinal_control_topic: "longitudinal_control"
    reset_topic: "/reset"
    desired_distance_topic: "desired_distance"
    convoy_control_data_topic: "convoy_control_data"
    shared_long_traj_topic: "shared_traj"
    virtual_leader_traj_topic: "/virtual_leader_traj"
    virtual_leader_frame: "virtual_link"

    # vehicle parameters/constraints
    base_lidar_distance: .270  # m
    vehicle_length: .560       # m
    wheelbase: .324            # m
    speed_min: -0.1            # m/s
    speed_max: 4.0             # m/s
    accel_min: -5.0            # m/s^2
    accel_max: 5.0             # m/s^2
    steer_min: -0.36           # rad
    steer_max: 0.36            # rad
    steer_v_min: -3.14         # rad/s
    steer_v_max: 3.14          # rad/s
    dt: 0.1                    # discretization time (s)
    tau: 0.1                   # accel delay time; tau=dt -> one timestep delay
    desired_distance: 1.0      # m

    # data logging
    learn_samples_cut: 10000

    ########## convoy_safety ########## 

    # autonomous_control
    safety_button_ind: 5  # R1
    brake_button_ind: 4   # L1

    ########## convoy_planning ########## 

    planner_trajectory_topic: 'planner_trajectory'
    planner_trajectory_vis_topic: 'planner_trajectory_vis'

    ########## convoy_perception ########## 

    # preprocess_lidar
    fov_deg: 270.0

    # course_walls_seg
    track_width: 2.5
    lidar_range_threshold: 10.0  # m
    dbscan_eps: .4
    dbscan_min_samples: 5
    min_cluster_size: 60
    neighbor_ball_radius: 0.7

    ########## convoy_control ########## 
    
    # random reference trajectory params
    stop_time: 30.0
    wait_time: 2.0
    brake_decel: -0.5
    leader_trajectory_type: 'stagger'  # random, stagger, or sinusoid
    zero_speed_percentage: 0.0
    stagger_time_min: 20.0
    stagger_time_max: 20.0
    sinusoid_period_min: 20.0
    sinusoid_period_max: 20.0
    low_speed_min: 2.0
    low_speed_max: 2.0
    high_speed_min: 3.5
    high_speed_max: 3.5
    reference_accel_min: 0.5
    reference_accel_max: 3.0

    ## longitudinal control ##

    # constant_input
    constant_speed: 0.0
    constant_accel: 0.0

    # double_int_mpc
    Q_double_int_weight: 10.0
    R_double_int_weight: 1.0

    # linear_feedback
    relative_position_gain_accel: 0.0
    relative_speed_gain_accel: 0.0
    relative_position_gain_vel: 1.0
    relative_speed_gain_vel: 2.0

    # DMPC
    horizon: 100

    # quad_pfmpc 
    q_self_quad_position: 1.
    q_self_quad_speed: 1.
    q_pred_quad_position: 1.
    q_pred_quad_speed: 1.
    r_pfmpc: .1

    # l1_pfmpc
    W_l1: [1., 1.]
    q_self_l1: 1.
    q_pred_l1: 1.
    r_l1: 1.

    # linear_mpc
    q_mpc: 10.
    q_f_mpc: 1.
    r_mpc: 0.1 

    ## lateral control ##

    # pure_pursuit
    lookahead_speed_gain: 0.75
    min_lookahead: 0.5
    max_lookahead: 2.0

    # stanley
    crosstrack_error_gain: 0.5
    softness_gain: 1.0

    ## bicycle model ##

    # bicycle_mpc
    Q_bicycle_weight: 3.0
    R_bicycle_weight: 1.0
